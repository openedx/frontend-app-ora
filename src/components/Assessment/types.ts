import { CriterionConfig, MutationStatus } from "data/services/lms/types";

export type Criterion = {
  optionsValue: string | null;
  optionsIsInvalid: boolean;
  optionsOnChange: (e: React.ChangeEvent<HTMLInputElement>) => void;

  feedbackValue: string | null;
  feedbackIsInvalid: boolean;
  feedbackOnChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
} & CriterionConfig;

export type RubricHookData = {
  criteria: Criterion[];
  onSubmit: () => void;
  submitStatus: MutationStatus;

  // overall feedback
  overallFeedback: string;
  onOverallFeedbackChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  overallFeedbackDisabled: boolean;
  overallFeedbackIsInvalid: boolean;
  overallFeedbackPrompt: string;
};