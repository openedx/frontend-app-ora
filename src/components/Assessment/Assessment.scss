@import "~@edx/brand/paragon/variables";
@import "~@edx/paragon/scss/core/core";
@import "~@edx/brand/paragon/overrides";

.criteria-label {
  width: 100%;
  .criteria-title {
    display: inline-block;
    max-width: calc(100% - 44px);
    color: $primary-500;
    font-weight: bold;
    vertical-align: top;
  }
  .esg-help-icon {
    float: right;
    margin-top: (map-get($spacers, 2) * -1);
    margin-right: (map-get($spacers, 2\.5) * -1);
    vertical-align: top;
  }
}
.criteria-option {
  width: 100%;
  > div {
    display: inline;
    width: 100%;
    .pgn__form-label {
      display: inline-flex;
    }
    .pgn__form-control-description {
      float: right;
    }
  }
}

.criterion-feedback {
  margin-top: 1rem;
}

.popover.overlay-help-popover {
  z-index: 4000;
  margin-right: map-get($spacers, 1) !important;
  .help-popover-option {
    margin-bottom: map-get($spacers, 1);
  }
}


.assessment-card {
  width: 320px !important;
  height: fit-content;
  max-height: 100%;
  margin-left: map-get($spacers, 3);
  position: sticky !important;
  top: map-get($spacers, 1) * -1;

  .assessment-header {
    box-shadow: 0 0 0.25rem rgba(0, 0, 0, 0.3) !important;
    display: flex;
    justify-content: center;
    padding: map-get($spacers, 3);
  }

  .assessment-body {
    overflow-y: scroll;
  }

  .assessment-footer {
    box-shadow: 0 0 0.25rem rgba(0, 0, 0, 0.3) !important;
    justify-content: center;
    padding: map-get($spacers, 3);
  }

  button.pgn__stateful-btn.pgn__stateful-btn-state-pending {
    opacity: .4 !important;
  }
}

@include media-breakpoint-down(sm) {
  .assessment-card {
    margin-left: 0 !important;
    width: 100% !important;
  } 
}
