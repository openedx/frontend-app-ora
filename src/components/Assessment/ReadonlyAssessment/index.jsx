import React from 'react';
import PropTypes from 'prop-types';

import { useHasSubmitted, useRefreshPageData } from 'hooks/app';
import { useSubmittedAssessment } from 'hooks/assessment';
import ReadOnlyAssessment from './ReadOnlyAssessment';

/*
 * If called manually, will include step data and manually passed assessment(s).
 * If called as part of Assessment base component, will display the submitted assessment.
 */
const ReadOnlyAssessmentContainer = (props) => {
  const hasSubmitted = useHasSubmitted();
  const submittedAssessment = useSubmittedAssessment();
  const refreshPageData = useRefreshPageData(hasSubmitted);
  // refresh page data on transfer to Read-only mode from editable
  React.useEffect(() => {
    if (hasSubmitted) {
      refreshPageData();
    }
  }, [hasSubmitted]); // eslint-disable-line react-hooks/exhaustive-deps
  return (
    <ReadOnlyAssessment
      {...props}
      {...hasSubmitted && { assessment: submittedAssessment }}
    />
  );
};
ReadOnlyAssessmentContainer.defaultProps = {
  defaultOpen: false,
  assessment: null,
  assessments: null,
  stepScore: null,
  stepLabel: null,
  step: null,
};
ReadOnlyAssessmentContainer.propTypes = {
  stepLabel: PropTypes.string,
  step: PropTypes.string,
  stepScore: PropTypes.shape({
    earned: PropTypes.number,
    total: PropTypes.number,
  }),
  defaultOpen: PropTypes.bool,
  assessment: PropTypes.shape({}),
  assessments: PropTypes.arrayOf(PropTypes.shape({})),
};

export default ReadOnlyAssessmentContainer;
