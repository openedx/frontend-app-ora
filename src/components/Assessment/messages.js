import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  gradeSubmitted: {
    id: 'frontend-app-ora.EditableAssessment.gradeSubmitted',
    defaultMessage: 'Grade Submitted',
    description: 'Submit Grade button text after successful submission',
  },
  rubric: {
    id: 'frontend-app-ora.EditableAssessment.rubric',
    defaultMessage: 'Rubric',
    description: 'Rubric interface label',
  },
  submitGrade: {
    id: 'frontend-app-ora.EditableAssessment.submitGrade',
    defaultMessage: 'Submit grade',
    description: 'Submit Grade button text',
  },
  submittingGrade: {
    id: 'frontend-app-ora.EditableAssessment.submittingGrade',
    defaultMessage: 'Submitting grade',
    description: 'Submit Grade button text while submitting',
  },
  overallComments: {
    id: 'frontend-app-ora.EditableAssessment.overallComments',
    defaultMessage: 'Overall comments',
    description: 'Rubric overall comments label',
  },
  addComments: {
    id: 'frontend-app-ora.EditableAssessment.addComments',
    defaultMessage: 'Add comments (Optional)',
    description: 'Rubric comments input label',
  },
  comments: {
    id: 'frontend-app-ora.EditableAssessment.comments',
    defaultMessage: 'Comments (Optional)',
    description: 'Rubric comments display label',
  },
  overallFeedbackError: {
    id: 'frontend-app-ora.EditableAssessment.overallFeedbackError',
    defaultMessage: 'The overall feedback is required',
    description: 'Error message when feedback input is required',
  },
  finishLater: {
    id: 'frontend-app-ora.EditableAssessment.finishLater',
    defaultMessage: 'Finish grading later',
    description: 'Button text for close modal action in the assessment view',
  },
});

export default messages;
