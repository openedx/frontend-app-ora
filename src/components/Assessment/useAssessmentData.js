import React from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import { useHasSubmitted, useInitializeAssessment } from 'hooks/assessment';

export const stateKeys = StrictDict({
  initialized: 'initialized',
});

export const useAssessmentData = () => {
  const [initialized, setInitialized] = useKeyedState(stateKeys.initialized, false);
  const hasSubmitted = useHasSubmitted();
  const initialize = useInitializeAssessment();
  React.useEffect(() => {
    initialize();
    setInitialized(true);
  }, [initialize, setInitialized]);
  return { initialized, hasSubmitted };
};

export default useAssessmentData;
