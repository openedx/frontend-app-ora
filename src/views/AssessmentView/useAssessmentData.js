import React from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import {
  useIsORAConfigLoaded,
  useIsPageDataLoaded,
  usePrompts,
  useResponse,
  useSetResponse,
  useResponseData,
} from 'hooks/app';

export const stateKeys = StrictDict({
  initialized: 'initialized',
});

const useAssessmentData = () => {
  const [initialized, setInitialized] = useKeyedState(stateKeys.initialized, false);
  const prompts = usePrompts();
  const response = useResponse();
  const responseData = useResponseData();
  const setResponse = useSetResponse();
  const isLoaded = useIsORAConfigLoaded();
  const isPageDataLoaded = useIsPageDataLoaded();
  React.useEffect(() => {
    if (!initialized && isLoaded && isPageDataLoaded) {
      setResponse(responseData);
      setInitialized(true);
    } else if (initialized && responseData && response !== responseData) {
      setResponse(responseData);
    }
  }, [responseData, initialized, isLoaded, isPageDataLoaded, response, setResponse, setInitialized]);
  return {
    isLoaded,
    response,
    prompts,
  };
};

export default useAssessmentData;
