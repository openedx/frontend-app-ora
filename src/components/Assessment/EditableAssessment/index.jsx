import React from 'react';

import { Card } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { useCriteriaConfig } from 'hooks/assessment';

import CriterionContainer from 'components/CriterionContainer';
import RadioCriterion from 'components/CriterionContainer/RadioCriterion';
import CriterionFeedback from 'components/CriterionContainer/CriterionFeedback';

import OverallFeedback from './OverallFeedback';
import AssessmentActions from './AssessmentActions';

import messages from '../messages';

/**
 * <Rubric />
 */
const EditableAssessment = () => {
  const { formatMessage } = useIntl();

  const criteria = useCriteriaConfig();

  const criteriaContainers = criteria.map(
    (criterion, criterionIndex) => (
      <CriterionContainer
        key={criterion.name}
        criterion={criterion}
        input={<RadioCriterion {...{ criterion, criterionIndex }} />}
        feedback={<CriterionFeedback {...{ criterion, criterionIndex }} />}
      />
    ),
  );

  return (
    <Card className="assessment-card">
      <Card.Section className="assessment-body">
        <h3>{formatMessage(messages.rubric)}</h3>
        <hr className="m-2.5" />
        {criteriaContainers}
        <hr />
        <OverallFeedback />
      </Card.Section>
      <AssessmentActions />
    </Card>
  );
};

EditableAssessment.propTypes = {
};

export default EditableAssessment;
