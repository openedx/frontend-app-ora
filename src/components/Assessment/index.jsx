import React from 'react';

import EditableAssessment from './EditableAssessment';
import ReadonlyAssessment from './ReadonlyAssessment';

import { useAssessmentData } from './useAssessmentData';

import './Assessment.scss';

/**
 * <Assessment />
 */
const Assessment = () => {
  const { initialized, hasSubmitted } = useAssessmentData();
  if (!initialized) {
    return null;
  }
  return hasSubmitted
    ? <ReadonlyAssessment defaultOpen />
    : <EditableAssessment />;
};
Assessment.propTypes = {};

export default Assessment;
