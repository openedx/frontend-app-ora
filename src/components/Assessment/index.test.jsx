import { shallow } from '@edx/react-unit-test-utils';

import Assessment from './index';

import { useAssessmentData } from './useAssessmentData';

jest.mock('./useAssessmentData', () => ({
  useAssessmentData: jest.fn(),
}));

jest.mock('./EditableAssessment', () => 'EditableAssessment');
jest.mock('./ReadonlyAssessment', () => 'ReadonlyAssessment');

describe('<Assessment />', () => {
  it('renders the ReadonlyAssessment', () => {
    useAssessmentData.mockReturnValue({ initialized: true, hasSubmitted: true });
    const wrapper = shallow(<Assessment />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('ReadonlyAssessment')).toHaveLength(1);
    expect(wrapper.instance.findByType('EditableAssessment')).toHaveLength(0);
  });
  it('renders the EditableAssessment', () => {
    useAssessmentData.mockReturnValue({ initialized: true, hasSubmitted: false });
    const wrapper = shallow(<Assessment />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('ReadonlyAssessment')).toHaveLength(0);
    expect(wrapper.instance.findByType('EditableAssessment')).toHaveLength(1);
  });
  it('renders nothing if not initialized', () => {
    useAssessmentData.mockReturnValue({ initialized: false });
    const wrapper = shallow(<Assessment />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.isEmptyRender()).toBe(true);
  });
});
