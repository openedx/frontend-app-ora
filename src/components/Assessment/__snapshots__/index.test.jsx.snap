// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Assessment /> renders nothing if not initialized 1`] = `null`;

exports[`<Assessment /> renders the EditableAssessment 1`] = `<EditableAssessment />`;

exports[`<Assessment /> renders the ReadonlyAssessment 1`] = `
<ReadonlyAssessment
  defaultOpen={true}
/>
`;
